cmd/webhook/admission/admission.go

```diff
@@ -0,0 +1,218 @@
+}
+
+type RouteGroupAdmitter struct {
+}
+
+func init() {
+	prometheus.MustRegister(totalRequests, invalidRequests, rejectedRequests, approvedRequests, admissionDuration)
+}
+
+func (r RouteGroupAdmitter) Name() string {
+	return "routegroup"
+}
+
+func (r RouteGroupAdmitter) Admit(req *admissionsv1.AdmissionRequest) (*admissionsv1.AdmissionResponse, error) {
+	rgItem := definitions.RouteGroupItem{}
+	err := json.Unmarshal(req.Object.Raw, &rgItem)
+	if err != nil {
+		emsg := fmt.Errorf("could not parse RouteGroup, %w", err)
+		log.Error(emsg)
+		return &admissionsv1.AdmissionResponse{
+			UID:     req.UID,
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: emsg.Error(),
+			},
+		}, nil
+	}
+
+	err = definitions.ValidateRouteGroup(&rgItem)
+	if err != nil {
+		emsg := fmt.Errorf("could not validate RouteGroup, %w", err)
+		log.Error(emsg)
+		return &admissionsv1.AdmissionResponse{
+			UID:     req.UID,
+			Allowed: false,
+			Result: &metav1.Status{
+				Message: emsg.Error(),
+			},
+		}, nil
+	}
+
+	return &admissionsv1.AdmissionResponse{
+		UID:     req.UID,
+		Allowed: true,
+	}, nil
+}
+
+func Handler(admitter Admitter) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		admitterName := admitter.Name()
+		totalRequests.WithLabelValues(admitterName).Inc()
+
+		if r.Method != "POST" || r.Header.Get("Content-Type") != "application/json" {
+			invalidRequests.WithLabelValues(admitterName).Inc()
+			w.WriteHeader(http.StatusBadRequest)
+		}
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			log.Errorf("Unable to read request: %v", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			invalidRequests.WithLabelValues(admitterName).Inc()
+			return
+		}
+
+		review := admissionsv1.AdmissionReview{}
+		err = json.Unmarshal(body, &review)
+		if err != nil {
+			log.Errorf("Unable to parse request: %v", err)
+			w.WriteHeader(http.StatusBadRequest)
+			invalidRequests.WithLabelValues(admitterName).Inc()
+			return
+		}
+
+		operationInfo := fmt.Sprintf(
+			"%s %s %s/%s",
+			review.Request.Operation,
+			review.Request.Kind,
+			review.Request.Namespace,
+			extractName(review.Request),
+		)
+
+		gvr := review.Request.Resource
+		group := gvr.Group
+		if group == "" {
+			group = "zalando.org"
+		}
+
+		labelValues := prometheus.Labels{
+			"admitter":     admitterName,
+			"operation":    string(review.Request.Operation),
+			"group":        group,
+			"version":      gvr.Version,
+			"resource":     gvr.Resource,
+			"sub_resource": review.Request.SubResource,
+		}
+
+		start := time.Now()
+		defer admissionDuration.With(labelValues).
+			Observe(float64(time.Since(start)) / float64(time.Second))
+
+		admResp, err := admitter.Admit(review.Request)
+		if err != nil {
+			log.Errorf("Rejected %s: %v", operationInfo, err)
+			writeResponse(w, errorResponse(review.Request.UID, err))
+			rejectedRequests.With(labelValues).Inc()
+			return
+		}
+
+		log.Infof("Allowed %s", operationInfo)
```

Comment:
I would consider if it is enough to log the successful admissions only with Debug, to avoid log spam